tests/unit/lms/views/exceptions_test.py

```diff
@@ -1,112 +1,138 @@
+from pyramid import httpexceptions
 
 from lms.models import ReusedConsumerKey
+from lms.resources import LTILaunchResource
 from lms.resources._js_config import JSConfig
 from lms.services import HAPIError
 from lms.validation import ValidationError
-from lms.views import exceptions
+from lms.views.exceptions import ExceptionViews
 
 
-class ExceptionViewTest:
-    view = None
-    exception = None
+class TestNotFound:
+    def test_it(self, assert_response, exception_views):
+        assert_response(exception_views.notfound(), 404, message="Page not found")
 
-    expected_result = None
-    response_status = None
+    @pytest.fixture
+    def exception(self):
+        return httpexceptions.HTTPNotFound()
 
-    def handle(self, pyramid_request):
-        if self.exception is None:
-            return type(self).view(  # pylint:disable=not-callable
-                mock.sentinel.exception, pyramid_request
-            )
 
-        return type(self).view(  # pylint:disable=not-callable
-            self.exception, pyramid_request
+class TestForbidden:
+    def test_it(self, assert_response, exception_views):
+        assert_response(
+            exception_views.forbidden(),
+            403,
+            message="You're not authorized to view this page",
         )
 
-    def test_it_sets_response_status(self, pyramid_request):
-        self.handle(pyramid_request)
+    @pytest.fixture
+    def exception(self):
+        return httpexceptions.HTTPForbidden()
 
-        assert pyramid_request.response.status_int == self.response_status
 
-    def test_it_produces_the_expected_template_vars(self, pyramid_request):
-        result = self.handle(pyramid_request)
+class TestHTTPClientError:
+    def test_it(self, assert_response, exception, exception_views):
+        assert_response(
+            exception_views.http_client_error(),
+            status=exception.status_int,
+            message="http_client_error_test_explanation",
+        )
 
-        assert result == self.expected_result
+    @pytest.fixture
+    def exception(self):
+        return httpexceptions.HTTPBadRequest("http_client_error_test_explanation")
 
 
-class TestNotFound(ExceptionViewTest):
-    view = exceptions.notfound
-    exception = None
+class TestHAPIError:
+    def test_it(self, assert_response, exception_views):
+        assert_response(
+            exception_views.hapi_error(), 500, message="hapi_error_test_explanation"
+        )
 
-    response_status = 404
-    expected_result = {"message": "Page not found"}
+    @pytest.fixture
+    def exception(self):
+        return HAPIError("hapi_error_test_explanation")
 
 
-class TestForbidden(ExceptionViewTest):
-    view = exceptions.forbidden
-    exception = None
+class TestValidationError:
+    def test_it(self, assert_response, exception, exception_views):
+        assert_response(
+            exception_views.validation_error(),
+            status=exception.status_int,
```

Comment:
I'm not sure why we specify this by reference here, where it's a concrete value elsewhere? I think this can only actually take one value (422?). It would make for nice documentation here to be able to see that.

Reply:
The actual behaviour of the exception view is that it sets the status to "whatever the `status_int` of the exception is" not to `422`. The `422` status is actually decided by the `ValidationError` exception class (because it subclasses `HTTPUnprocessableEntity` which has `status_int` `422`) :confused:

There could be another `ValidationError` subclass with a different `status_int` that would get caught by this exception view and result in a different status. (AFAIK there isn't, but there could be.)

I think maybe `ValidationError` shouldn't subclass `HTTPUnprocessableEntity` and shouldn't have a `status_int`, and it should be up to the exception view to decide what status to use, at which point this test would say `status=422` (or whatever status we chose to use).

On the other hand the `http_client_error()` exception view that catches Pyramid `HTTPClientError`'s also does the same thing with setting the status to whatever the exception's `status_int` is :man_shrugging: